### ISSUE TITLE
[Nametag] Send chat messages for full self disclosure

### ISSUE THREAD

**Issue by Shaomei** [aimpowered/LibOrate#73]
See figma design
<img width="322" alt="Screenshot 2024-06-04 at 12 24 42 PM" src="https://github.com/aimpowered/ZoomCompanion/assets/202047/7e7ec136-4bc4-4bc2-aeb8-5ef4ece9297a">

**Comment by Shaomei** [comment]
The chat message will be sent to all participants when the user "submit" the nametag.

**Comment by Shaomei** [comment]
Extension: while the nametag is on, send private message to everyone who joins after the initial public message.

**Comment by BrennanJones** [comment]
Hi @Shaomei and @blickly , an update on this: I was able to mostly figure out the OAuth flow. This involves fetching the OAuth authorization code (which must be done at the moment the user installs the app) and using this to fetch and save the user's access and refresh tokens, which are used for server-side API calls to the Zoom API. The access token, which is unique to each user, is used for the API calls themselves, and lasts for one hour. The refresh token lasts for 90 days, and is used to fetch a new access token and refresh token. In theory, if the user does not use the app for longer than 90 days, then both tokens expire -- at which point, we have to go through the whole cycle again to get new tokens (the user might need to reinstall the app again at this point, in order to give the app a new authorization code for fetching access and refresh tokens).

However, even though I figured out this OAuth flow, I realized that the API call for sending a message to the chat (`POST /chat/users/{userId}/messages`: https://developers.zoom.us/docs/api/rest/reference/chat/methods/#operation/sendaChatMessage) only works for continuous team chats (i.e., the Zoom Workplace Team Chats that persist after meetings, not the normal meeting chat). It turns out that, even for the server-side API, there is no API call for automatically sending a message to the meeting chat (at least according to this: https://devforum.zoom.us/t/use-api-to-send-chat-messages-to-zoom-meeting/96951#:~:text=Oct%202023-,%40prodoer.matp9%20The%20send%20chat%20message%20API%20is%20used%20to%20send%20chat%20messages%20in%20%E2%80%9CZoom%20Teams%20Chat%E2%80%9D.%20We%20do%20not%20have%20apis%20to%20send%20messages%20in%20live%20meetings,-Solution). In other words, this does not seem to be possible with Zoom's current APIs.

**Comment by BrennanJones** [comment]
> Hi @Shaomei and @blickly , an update on this: I was able to mostly figure out the OAuth flow. This involves fetching the OAuth authorization code (which must be done at the moment the user installs the app) and using this to fetch and save the user's access and refresh tokens, which are used for server-side API calls to the Zoom API. The access token, which is unique to each user, is used for the API calls themselves, and lasts for one hour. The refresh token lasts for 90 days, and is used to fetch a new access token and refresh token. In theory, if the user does not use the app for longer than 90 days, then both tokens expire -- at which point, we have to go through the whole cycle again to get new tokens (the user might need to reinstall the app again at this point, in order to give the app a new authorization code for fetching access and refresh tokens). I successfully got the OAuth flow working on my test deployment and used it to successfully call server-side API methods, including `POST /users/me` to get the user's information.
> 
> However, even though I figured out this OAuth flow, I realized that the API call for sending a message to the chat (`POST /chat/users/{userId}/messages`: https://developers.zoom.us/docs/api/rest/reference/chat/methods/#operation/sendaChatMessage) only works for continuous team chats (i.e., the Zoom Workplace Team Chats that persist after meetings, not the normal meeting chat). It turns out that, even for the server-side API, there is no API call for automatically sending a message to the meeting chat (at least according to this: https://devforum.zoom.us/t/use-api-to-send-chat-messages-to-zoom-meeting/96951#:~:text=Oct%202023-,%40prodoer.matp9%20The%20send%20chat%20message%20API%20is%20used%20to%20send%20chat%20messages%20in%20%E2%80%9CZoom%20Teams%20Chat%E2%80%9D.%20We%20do%20not%20have%20apis%20to%20send%20messages%20in%20live%20meetings,-Solution). In other words, this does not seem to be possible with Zoom's current APIs. I tried using the `POST /chat/users/me/messages` server-side API call to send a chat message, but it requires a Workplace Team Chat channel ID to be passed into the `to_channel` parameter, and does not seem to work with normal Zoom meeting chats.

More info about the OAuth process: https://developers.zoom.us/docs/integrations/oauth/

**Comment by Shaomei** [comment]
Is there any way we can do this through a Zoom bot? I see that Otter.ai's
zoom bot is able to send messages.

**Comment by BrennanJones** [comment]
I found a potential solution that may allow us to accomplish what we want with this feature.

Neither the Zoom Apps SDK nor the Zoom server-side API allow us to send chat messages to meeting chats. The Zoom Meeting SDK, which is normally used for creating standalone apps that connect to Zoom meetings and handle audio/video streams and other data (not used for apps that run inside the Zoom client) allows us to do this. However, the process for doing this is normally complicated and requires setting up additional server-side infrastructure and so forth. Details: https://devforum.zoom.us/t/how-do-note-taking-bots-join-zoom-meetings/76040/2

However, there are third-party tools that allow us to more easily create bots that can join Zoom meetings in order to automatically send chat messages to the meeting participants (they can also be used for recording meetings and such, but we only need it for sending chat messages). One of such tools is Recall.ai (https://www.recall.ai/).

They have a getting-started page for creating a Zoom bot: https://docs.recall.ai/docs/set-up-zoom

As well as an API method for sending chat messages within the call: https://docs.recall.ai/docs/sending-chat-messages

There may be some limitations in that the bot has to join just as a participant normally would, and if the meeting the user is in has a wait room, the host of the meeting may need to allow the bot into the meeting. Also, some Zoom meetings require CAPTCHA verification to join, in order to prevent bots from joining (I'd assume this would also present bots from Otter.ai and other tools from joining too). But otherwise, I could see us having our LibOrate app send a command to our bot to join the meeting that the user is in once the user joins (by sending the bot a meeting ID and passcode), and then the bot would send a chat message once our Zoom app gives it a command to send the chat message.

There may be a cost to using Recall.ai. They have a pay-as-you-go plan which costs $1 per bot-hour of usage: https://docs.recall.ai/page/self-serve-plan-pay-as-you-go. Their site indicates that if we want to bring our app into production and switch to a monthly plan, to "simply reach out to support@recall.ai." My guess it that what they charge us might depend on the type and size of our organization, the number of users of our app, and potentially other things (in our case, because we're a non-profit, we might be able to get a lower monthly fee).

**Comment by BrennanJones** [comment]
Here's a sample meeting bot app built with Recall.ai: https://developers.zoom.us/blog/meetingbot-starter-kit-recallai/ (source code: https://github.com/zoom/meetingbot-recall-sample).

TL;DR: Using Recall.ai is likely the easiest and quickest solution, but it costs money (which we would have to consider when we eventually make our app available to the public). Recall.ai has a pay-as-you-go plan (which costs $1 per bot-hour of usage) and monthly plans (they don't state what the cost of their monthly plan is, only to "reach out to support@recall.ai" to switch to a monthly plan). My guess is they would likely decide how much to charge us per month based on the type of organization, our number of users, etc. (Since we're a non-profit, we _might_ be able to negotiate a lower monthly price, but I'm not sure.)

**Comment by BrennanJones** [comment]
I asked Recall.ai support about how the cost structure for the pay-as-you-go plan, and this is their response:

> Hey Brennan, thanks for reaching out!
> 
> Your high-level understanding is mostly correct, but there are some nuances so let me break this down a bit.
> 
> > As far as we understand, the pay-as-you-go plan costs $1 per bot-hour of usage. How is "bot-hour" measured? Is this the number of hours that each bot spends in meetings? E.g., if we have three bots that each spend one hour in meetings, is that three bot-hours?
> 
> You're correct that if you have 3 bots that each spend an hour in a meeting, this is 3 bot-hours.
> 
> A bot-hour is the time a bot is "alive", measured by its first in_call or in_waiting_room event, up until it it terminated (done).
> 
> > Also, is the cost rounded up to the nearest bot-hour? For example, if a bot spends only five minutes in a meeting, does that count as 0.08 bot-hours (and thus cost $0.08)? Or is this rounded up to one bot-hour (and thus $1.00)?
> 
> Bot usage is prorated to the nearest second. So if a bot is on a call for 10 seconds, this is only 10 bot-seconds of use and isn't rounded further than that.
> 
> Here's a doc that explains this a bit more with an example: https://docs.recall.ai/docs/bot-usage

In summary, they don't charge per hour of bot usage, but rather per _second_ of bot usage, at a rate of $1 USD / hour. Because our needs with the bot are minimal (just send a chat message to the meeting chat then leave the meeting), we can thus keep our costs with this API minimal.

We could consider using Recall.ai under their pay-as-you-go pricing scheme for the beta testing (since we'll have a minimal number of users), then consider whether we want to continue using their API (and switch to a monthly plan) or implement another solution when we release the app to the public.

**Comment by Shaomei** [comment]
Interesting! Thanks for finding out. Yeah let’s go with it for the beta
test, and we can figure something out later for scaling when Ben and I are
back.

Thank you for all the investigation, Brennan!!

**Comment by hotaery** [comment]
Hi @Shaomei. This issue is to add a full self-disclosure input box to allow users to enter full self-disclosure. When the user opens the Display Name Tag, a full self-disclosure will be sent to all meeting participants. If other people join the meeting later, a full self-disclosure will be sent privately. Am I right?
